firedrake/cython/dmcommon.pyx

```diff
@@ -666,6 +668,174 @@ def quadrilateral_closure_ordering(PETSc.DM plex,
+                                              PetscInt eStart,
+                                              PetscInt eEnd,
+                                              PetscInt *cone,
+                                              PetscInt coneSize):
+    cdef:
+        PetscInt k, n, e, q, o = 0, coneSize1 = coneSize
+        PetscInt *cone1 = NULL
+        PetscInt *inds = NULL
+
+    CHKERR(PetscMalloc1(coneSize, &cone1))
+    CHKERR(PetscMalloc1(coneSize, &inds))
+    for k in range(coneSize1):
+        cone1[k] = cone[k]
+    n = 0
+    for e in range(eStart, eEnd):
+        q = cell_closure[cell, e]
+        for k in range(coneSize1):
+            if q == cone1[k]:
+                inds[n] = k
+                n += 1
+                break
+        else:
+            continue
+        while k < coneSize1 - 1:
+            cone1[k] = cone1[k + 1]
+            k += 1
+        coneSize1 -= 1
+    assert n == coneSize
+    for k in range(n):
+        o += np.math.factorial(n - 1 - k) * inds[k]
+    CHKERR(PetscFree(cone1))
+    CHKERR(PetscFree(inds))
+    return o
+
+
+cdef inline PetscInt _ncr(PetscInt n , PetscInt r):
+    return np.math.factorial(n) // np.math.factorial(r) // np.math.factorial(n - r)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def entity_orientations(mesh,
+                        np.ndarray[PetscInt, ndim=2, mode="c"] cell_closure):
+    """Compute entity orientations.
+
+    :arg mesh: The `MeshTopology` object encapsulating the mesh topology
+    :arg cell_closure: The two-dimensional array, each row of which contains
+        the closure of the associated cell
+
+    Each row of cell_closure must have been ordered to work well with
+    FInAT entity_dofs to create cell_node map (see `get_cell_nodes`).
+    """
+    cdef:
+        PETSc.DM dm
+        PetscInt dim, d, p, cell, e, eStart, eEnd, numCells, o
+        PetscInt *cone = NULL, c2, c3
+        PetscInt coneSize
+        PetscInt *ecounts = NULL, ecount
+        PetscInt *eoffsets = NULL, eoffset
+        PetscInt *inds = NULL
+        np.ndarray[PetscInt, ndim=2, mode="c"] entity_orientations
+
+    if type(mesh) is not firedrake.mesh.MeshTopology:
+        raise TypeError(f"Unexpected mesh type: {type(mesh)}")
+    dm = mesh.topology_dm
+    ufl_cell = mesh.ufl_cell()
+    dim = ufl_cell.topological_dimension()
+    numCells = cell_closure.shape[0]
+    entity_orientations = np.zeros_like(cell_closure)
+    if ufl_cell.is_simplex():
+        # FInAT reference triangle:
+        #
+        #   1
+        #   | \
+        #   5   3
+        #   | 6   \
+        #   0--4---2
+        #
+        CHKERR(PetscMalloc1(dim + 1, &ecounts))
+        CHKERR(PetscMalloc1(dim + 2, &eoffsets))
+        for d in range(dim + 1):
+            ecounts[d] = _ncr(dim + 1, d + 1)
+        eoffsets[0] = 0
+        for d in range(1, dim + 2):
+            eoffsets[d] = eoffsets[d - 1] + ecounts[d - 1]
+        for cell in range(numCells):
```

Comment:
I think that this code needs a little exposition explaining what it is doing.

Reply:
Added more comments. Thanks!